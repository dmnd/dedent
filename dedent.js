// @flow

export default function dedent(
  strings: string | Array<string>,
  ...values: Array<string>
) {
  // $FlowFixMe: Flow doesn't undestand .raw
  const raw = typeof strings === "string" ? [strings] : strings.raw;

  // first, perform interpolation
  let result = "";
  for (let i = 0; i < raw.length; i++) {
    result += raw[i]
      // join lines when there is a suppressed newline
      .replace(/\\\n[ \t]*/g, "")
      // handle escaped backticks
      .replace(/\\`/g, "`");

    if (i < values.length) {
      result += values[i];
    }
  }

  // now strip indentation
  const lines = result.split("\n");
  let mindent: number | null = null;
  lines.forEach(l => {
    let m = l.match(/^(\s+)\S+/);
    if (m) {
      let indent = m[1].length;
      if (!mindent) {
        // this is the first indented line
        mindent = indent;
      } else {
        mindent = Math.min(mindent, indent);
      }
    }
  });

  if (mindent !== null) {
    const m = mindent; // appease Flow
    result = lines.map(l => l[0] === " " ? l.slice(m) : l).join("\n");
  }

  if (result.startsWith('\n')) {
    result = result.substr(1);
  }
  return result.replace(/^((.|\n)*)\n[ \t]*?$/, '$1');
}
