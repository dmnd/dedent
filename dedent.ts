export default function dedent(literals: string): string;
export default function dedent(
  strings: TemplateStringsArray,
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ...values: any[]
): string;
export default function dedent(
  strings: string | TemplateStringsArray,
  ...values: string[]
) {
  const raw = typeof strings === "string" ? [strings] : strings.raw;

  // first, perform interpolation
  let result = "";
  for (let i = 0; i < raw.length; i++) {
    result += raw[i]
      // handle escaped newlines, backticks, and interpolation characters
      .replace(/\\`/g, "`")
      .replace(/\\\n[ \t]*/g, "")
      .replace(/\\\$/g, "$")
      .replace(/\\{/g, "{");

    if (i < values.length) {
      result += values[i];
    }
  }

  // now strip indentation
  const lines = result.split("\n");
  let mindent: number | null = null;
  for (const l of lines) {
    const m = l.match(/^(\s+)\S+/);
    if (m) {
      const indent = m[1].length;
      if (!mindent) {
        // this is the first indented line
        mindent = indent;
      } else {
        mindent = Math.min(mindent, indent);
      }
    }
  }

  if (mindent !== null) {
    const m = mindent; // appease TypeScript
    result = lines
      // https://github.com/typescript-eslint/typescript-eslint/issues/7140
      // eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
      .map((l) => (l[0] === " " || l[0] === "\t" ? l.slice(m) : l))
      .join("\n");
  }

  return (
    result
      // dedent eats leading and trailing whitespace too
      .trim()
      // handle escaped newlines at the end to ensure they don't get stripped too
      .replace(/\\n/g, "\n")
  );
}
