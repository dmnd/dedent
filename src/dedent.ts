import type { Dedent, DedentOptions } from "./types.js";

export type * from "./types.js";

const dedent: Dedent = createDedent({});

export default dedent;

function createDedent(options: DedentOptions) {
	dedent.withOptions = (newOptions: DedentOptions): Dedent =>
		createDedent({ ...options, ...newOptions });

	return dedent;

	function dedent(literals: string): string;
	function dedent(strings: TemplateStringsArray, ...values: unknown[]): string;
	function dedent(
		strings: TemplateStringsArray | string,
		...values: unknown[]
	) {
		const raw = typeof strings === "string" ? [strings] : strings.raw;
		const { escapeSpecialCharacters = Array.isArray(strings) } = options;

		// first, perform interpolation
		let result = "";
		for (let i = 0; i < raw.length; i++) {
			let next = raw[i];

			if (escapeSpecialCharacters) {
				// handle escaped newlines, backticks, and interpolation characters
				next = next
					.replace(/\\\n[ \t]*/g, "")
					.replace(/\\`/g, "`")
					.replace(/\\\$/g, "$")
					.replace(/\\\{/g, "{");
			}

			result += next;

			if (i < values.length) {
				// eslint-disable-next-line @typescript-eslint/restrict-plus-operands
				result += values[i];
			}
		}

		// now strip indentation
		const lines = result.split("\n");
		let mindent: null | number = null;
		for (const l of lines) {
			const m = l.match(/^(\s+)\S+/);
			if (m) {
				const indent = m[1].length;
				if (!mindent) {
					// this is the first indented line
					mindent = indent;
				} else {
					mindent = Math.min(mindent, indent);
				}
			}
		}

		if (mindent !== null) {
			const m = mindent; // appease TypeScript
			result = lines
				// https://github.com/typescript-eslint/typescript-eslint/issues/7140
				// eslint-disable-next-line @typescript-eslint/prefer-string-starts-ends-with
				.map((l) => (l[0] === " " || l[0] === "\t" ? l.slice(m) : l))
				.join("\n");
		}

		return (
			result
				// dedent eats leading and trailing whitespace too
				.trim()
				// handle escaped newlines at the end to ensure they don't get stripped too
				.replace(/\\n/g, "\n")
		);
	}
}
