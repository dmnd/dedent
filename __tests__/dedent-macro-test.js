import pluginTester from "babel-plugin-tester";
import plugin from "babel-plugin-macros";

pluginTester({
  plugin,
  snapshot: true,
  babelOptions: { filename: __filename },
  tests: [
    {
      title: "as a function",
      code: `
        import dedent from "../macro";

        dedent(\`
          some stuff
        \`);
      `
    },
    {
      title: "as a template string",
      code: `
        import dedent from "../macro";

        dedent\`
          some stuff
        \`;
      `
    },
    {
      title: "multiple indentations",
      code: `
        import dedent from "../macro";

        dedent\`
            first
              second
                  third
        \`;
      `
    },
    {
      title: "expressions",
      code: `
        import dedent from "../macro";

        dedent\`
          first ${"line"}
            ${"second"}
            third
        \`;
      `
    }
  ]
});
