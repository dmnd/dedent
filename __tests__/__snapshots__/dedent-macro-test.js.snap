exports[`as a function 1`] = `
"
import dedent from \"../macro\";

dedent(\`
  some stuff
\`);

      ↓ ↓ ↓ ↓ ↓ ↓

\"some stuff\";
"
`;

exports[`as a template string 1`] = `
"
import dedent from \"../macro\";

dedent\`
  some stuff
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

\"some stuff\";
"
`;

exports[`explicit newline 1`] = `
"
import dedent from \"../macro\";

dedent\`
  <p>
    Hello world!
  </p>

\`;

      ↓ ↓ ↓ ↓ ↓ ↓

\"<p>\\n  Hello world!\\n</p>\";
"
`;

exports[`expressions 1`] = `
"
import dedent from \"../macro\";

dedent\`
  first line
    second
    third
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

\"first line\\n  second\\n  third\";
"
`;

exports[`multiple indentations 1`] = `
"
import dedent from \"../macro\";

dedent\`
    first
      second
          third
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

\"first\\n  second\\n      third\";
"
`;
